File: ts/src/directMessageSubscriptions.ts
import type { Filter } from "nostr-tools"

import { APP_KEYS_EVENT_KIND, INVITE_RESPONSE_KIND, MESSAGE_EVENT_KIND } from "./types"

export interface RegisteredDirectMessageSubscription {
  token: number
  addedAuthors: string[]
}

export interface RegisteredRuntimeSubscription {
  token: number
  addedAppKeysAuthors: string[]
  addedMessageAuthors: string[]
  addedInviteResponseRecipients: string[]
}

const normalizeAuthor = (value: unknown): string | null => {
  if (typeof value !== "string") return null
  const normalized = value.trim().toLowerCase()
  if (!/^[0-9a-f]{64}$/.test(normalized)) return null
  return normalized
}

export function directMessageSubscriptionAuthors(filter: Filter): string[] {
  if (!Array.isArray(filter.kinds) || !filter.kinds.includes(MESSAGE_EVENT_KIND)) {
    return []
  }

  if (!Array.isArray(filter.authors)) {
    return []
  }

  const authors: string[] = []
  const seen = new Set<string>()
  for (const author of filter.authors) {
    const normalized = normalizeAuthor(author)
    if (!normalized || seen.has(normalized)) continue
    seen.add(normalized)
    authors.push(normalized)
  }
  return authors
}

export function appKeysSubscriptionAuthors(filter: Filter): string[] {
  if (!Array.isArray(filter.kinds) || !filter.kinds.includes(APP_KEYS_EVENT_KIND)) {
    return []
  }

  if (!Array.isArray(filter.authors)) {
    return []
  }

  const authors: string[] = []
  const seen = new Set<string>()
  for (const author of filter.authors) {
    const normalized = normalizeAuthor(author)
    if (!normalized || seen.has(normalized)) continue
    seen.add(normalized)
    authors.push(normalized)
  }
  return authors
}

export function buildDirectMessageBackfillFilter(
  authors: Iterable<string>,
  since: number,
  limit: number = 200
): Filter {
  const normalizedAuthors: string[] = []
  const seen = new Set<string>()
  for (const author of authors) {
    const normalized = normalizeAuthor(author)
    if (!normalized || seen.has(normalized)) continue
    seen.add(normalized)
    normalizedAuthors.push(normalized)
  }

  return {
    kinds: [MESSAGE_EVENT_KIND],
    authors: normalizedAuthors,
    since,
    limit,
  }
}

export function buildAppKeysBackfillFilter(
  authors: Iterable<string>,
  limit: number = 200
): Filter {
  const normalizedAuthors: string[] = []
  const seen = new Set<string>()
  for (const author of authors) {
    const normalized = normalizeAuthor(author)
    if (!normalized || seen.has(normalized)) continue
    seen.add(normalized)
    normalizedAuthors.push(normalized)
  }

  return {
    kinds: [APP_KEYS_EVENT_KIND],
    authors: normalizedAuthors,
    limit,
  }
}

export function inviteResponseSubscriptionRecipients(filter: Filter): string[] {
  if (!Array.isArray(filter.kinds) || !filter.kinds.includes(INVITE_RESPONSE_KIND)) {
    return []
  }

  const recipients = (filter as Record<string, unknown>)["#p"]
  if (!Array.isArray(recipients)) {
    return []
  }

  const normalizedRecipients: string[] = []
  const seen = new Set<string>()
  for (const recipient of recipients) {
    const normalized = normalizeAuthor(recipient)
    if (!normalized || seen.has(normalized)) continue
    seen.add(normalized)
    normalizedRecipients.push(normalized)
  }
  return normalizedRecipients
}

export function buildInviteResponseBackfillFilter(
  recipients: Iterable<string>,
  since: number,
  limit: number = 200
): Filter {
  const normalizedRecipients: string[] = []
  const seen = new Set<string>()
  for (const recipient of recipients) {
    const normalized = normalizeAuthor(recipient)
    if (!normalized || seen.has(normalized)) continue
    seen.add(normalized)
    normalizedRecipients.push(normalized)
  }

  return {
    kinds: [INVITE_RESPONSE_KIND],
    "#p": normalizedRecipients,
    since,
    limit,
  }
}

export function buildRuntimeBackfillFilters(
  registered: Pick<
    RegisteredRuntimeSubscription,
    "addedAppKeysAuthors" | "addedMessageAuthors" | "addedInviteResponseRecipients"
  >,
  since: number,
  limit: number = 200
): Filter[] {
  const filters: Filter[] = []
  if (registered.addedAppKeysAuthors.length > 0) {
    filters.push(buildAppKeysBackfillFilter(registered.addedAppKeysAuthors, limit))
  }
  if (registered.addedMessageAuthors.length > 0) {
    filters.push(
      buildDirectMessageBackfillFilter(registered.addedMessageAuthors, since, limit)
    )
  }
  if (registered.addedInviteResponseRecipients.length > 0) {
    filters.push(
      buildInviteResponseBackfillFilter(
        registered.addedInviteResponseRecipients,
        since,
        limit
      )
    )
  }
  return filters
}

export class DirectMessageSubscriptionTracker {
  private nextToken = 1
  private authorsByToken = new Map<number, string[]>()
  private authorRefCounts = new Map<string, number>()

  registerFilter(filter: Filter): RegisteredDirectMessageSubscription {
    const token = this.nextToken++
    const authors = directMessageSubscriptionAuthors(filter)
    if (authors.length === 0) {
      return { token, addedAuthors: [] }
    }

    this.authorsByToken.set(token, authors)
    const addedAuthors: string[] = []
    for (const author of authors) {
      const refCount = this.authorRefCounts.get(author) || 0
      if (refCount === 0) {
        addedAuthors.push(author)
      }
      this.authorRefCounts.set(author, refCount + 1)
    }

    return { token, addedAuthors }
  }

  unregister(token: number): void {
    const authors = this.authorsByToken.get(token)
    if (!authors) return

    this.authorsByToken.delete(token)
    for (const author of authors) {
      const nextCount = Math.max((this.authorRefCounts.get(author) || 1) - 1, 0)
      if (nextCount === 0) {
        this.authorRefCounts.delete(author)
      } else {
        this.authorRefCounts.set(author, nextCount)
      }
    }
  }

  trackedAuthors(): string[] {
    return Array.from(this.authorRefCounts.keys()).sort()
  }
}

export class RuntimeSubscriptionTracker {
  private nextToken = 1
  private appKeysAuthorsByToken = new Map<number, string[]>()
  private messageAuthorsByToken = new Map<number, string[]>()
  private inviteResponseRecipientsByToken = new Map<number, string[]>()
  private appKeysAuthorRefCounts = new Map<string, number>()
  private messageAuthorRefCounts = new Map<string, number>()
  private inviteResponseRecipientRefCounts = new Map<string, number>()

  registerFilter(filter: Filter): RegisteredRuntimeSubscription {
    const token = this.nextToken++
    const addedAppKeysAuthors = registerValues(
      appKeysSubscriptionAuthors(filter),
      this.appKeysAuthorsByToken,
      this.appKeysAuthorRefCounts,
      token
    )
    const addedMessageAuthors = registerValues(
      directMessageSubscriptionAuthors(filter),
      this.messageAuthorsByToken,
      this.messageAuthorRefCounts,
      token
    )
    const addedInviteResponseRecipients = registerValues(
      inviteResponseSubscriptionRecipients(filter),
      this.inviteResponseRecipientsByToken,
      this.inviteResponseRecipientRefCounts,
      token
    )

    return {
      token,
      addedAppKeysAuthors,
      addedMessageAuthors,
      addedInviteResponseRecipients,
    }
  }

  unregister(token: number): void {
    unregisterValues(token, this.appKeysAuthorsByToken, this.appKeysAuthorRefCounts)
    unregisterValues(token, this.messageAuthorsByToken, this.messageAuthorRefCounts)
    unregisterValues(
      token,
      this.inviteResponseRecipientsByToken,
      this.inviteResponseRecipientRefCounts
    )
  }

  trackedMessageAuthors(): string[] {
    return Array.from(this.messageAuthorRefCounts.keys()).sort()
  }

  trackedAppKeysAuthors(): string[] {
    return Array.from(this.appKeysAuthorRefCounts.keys()).sort()
  }

  trackedInviteResponseRecipients(): string[] {
    return Array.from(this.inviteResponseRecipientRefCounts.keys()).sort()
  }
}

function registerValues(
  values: string[],
  valuesByToken: Map<number, string[]>,
  refCounts: Map<string, number>,
  token: number
): string[] {
  if (values.length === 0) {
    return []
  }

  valuesByToken.set(token, values)
  const addedValues: string[] = []
  for (const value of values) {
    const refCount = refCounts.get(value) || 0
    if (refCount === 0) {
      addedValues.push(value)
    }
    refCounts.set(value, refCount + 1)
  }
  return addedValues
}

function unregisterValues(
  token: number,
  valuesByToken: Map<number, string[]>,
  refCounts: Map<string, number>
): void {
  const values = valuesByToken.get(token)
  if (!values) return

  valuesByToken.delete(token)
  for (const value of values) {
    const nextCount = Math.max((refCounts.get(value) || 1) - 1, 0)
    if (nextCount === 0) {
      refCounts.delete(value)
    } else {
      refCounts.set(value, nextCount)
    }
  }
}

File: rust/crates/nostr-double-ratchet/src/direct_message_subscriptions.rs
use crate::{
    utils::pubkey_from_hex, APP_KEYS_EVENT_KIND, INVITE_RESPONSE_KIND, MESSAGE_EVENT_KIND,
};
use nostr::{Alphabet, Filter, Kind, PublicKey, SingleLetterTag, Timestamp};
use serde_json::Value;
use std::collections::{HashMap, HashSet};

const DIRECT_MESSAGE_RUNTIME_SUBSCRIPTION_ID: &str = "ndr-runtime-messages";
const DIRECT_MESSAGE_RUNTIME_SUBSCRIPTION_PREFIX: &str = "ndr-runtime-messages-";
const DIRECT_MESSAGE_PROTOCOL_SUBSCRIPTION_ID: &str = "icp-messages";
const DIRECT_MESSAGE_PROTOCOL_SUBSCRIPTION_PREFIX: &str = "icp-messages-";
const SESSION_CURRENT_SUBSCRIPTION_PREFIX: &str = "session-current-";
const SESSION_NEXT_SUBSCRIPTION_PREFIX: &str = "session-next-";
const INVITE_RESPONSE_SUBSCRIPTION_PREFIX: &str = "invite-responses-";

#[derive(Debug, Default, Clone)]
pub struct DirectMessageSubscriptionTracker {
    authors_by_subid: HashMap<String, Vec<PublicKey>>,
    author_ref_counts: HashMap<PublicKey, usize>,
}

impl DirectMessageSubscriptionTracker {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn register_subscription(
        &mut self,
        subid: impl AsRef<str>,
        filter_json: impl AsRef<str>,
    ) -> Vec<PublicKey> {
        let subid = subid.as_ref().trim();
        if subid.is_empty() {
            return Vec::new();
        }

        self.unregister_subscription(subid);

        let authors = direct_message_subscription_authors(subid, filter_json.as_ref());
        if authors.is_empty() {
            return Vec::new();
        }

        let mut added = Vec::new();
        for author in &authors {
            let ref_count = self.author_ref_counts.entry(*author).or_insert(0);
            if *ref_count == 0 {
                added.push(*author);
            }
            *ref_count += 1;
        }
        self.authors_by_subid.insert(subid.to_string(), authors);
        added.sort_by_key(|pubkey| pubkey.to_hex());
        added
    }

    pub fn unregister_subscription(&mut self, subid: impl AsRef<str>) {
        let subid = subid.as_ref().trim();
        if subid.is_empty() {
            return;
        }

        let Some(previous_authors) = self.authors_by_subid.remove(subid) else {
            return;
        };

        for author in previous_authors {
            let next_count = self
                .author_ref_counts
                .get(&author)
                .copied()
                .unwrap_or(1)
                .saturating_sub(1);
            if next_count == 0 {
                self.author_ref_counts.remove(&author);
            } else {
                self.author_ref_counts.insert(author, next_count);
            }
        }
    }

    pub fn tracked_authors(&self) -> Vec<PublicKey> {
        let mut authors: Vec<PublicKey> = self.author_ref_counts.keys().copied().collect();
        authors.sort_by_key(|pubkey| pubkey.to_hex());
        authors
    }
}

#[derive(Debug, Default, Clone, PartialEq, Eq)]
pub struct RuntimeSubscriptionRegistration {
    pub added_app_keys_authors: Vec<PublicKey>,
    pub added_message_authors: Vec<PublicKey>,
    pub added_invite_response_recipients: Vec<PublicKey>,
}

#[derive(Debug, Default, Clone)]
pub struct RuntimeSubscriptionTracker {
    app_keys_authors_by_subid: HashMap<String, Vec<PublicKey>>,
    app_keys_author_ref_counts: HashMap<PublicKey, usize>,
    message_authors_by_subid: HashMap<String, Vec<PublicKey>>,
    message_author_ref_counts: HashMap<PublicKey, usize>,
    invite_response_recipients_by_subid: HashMap<String, Vec<PublicKey>>,
    invite_response_recipient_ref_counts: HashMap<PublicKey, usize>,
}

impl RuntimeSubscriptionTracker {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn register_subscription(
        &mut self,
        subid: impl AsRef<str>,
        filter_json: impl AsRef<str>,
    ) -> RuntimeSubscriptionRegistration {
        let subid = subid.as_ref().trim();
        if subid.is_empty() {
            return RuntimeSubscriptionRegistration::default();
        }

        self.unregister_subscription(subid);

        let added_app_keys_authors = register_pubkeys(
            &mut self.app_keys_authors_by_subid,
            &mut self.app_keys_author_ref_counts,
            subid,
            app_keys_subscription_authors(subid, filter_json.as_ref()),
        );
        let added_message_authors = register_pubkeys(
            &mut self.message_authors_by_subid,
            &mut self.message_author_ref_counts,
            subid,
            direct_message_subscription_authors(subid, filter_json.as_ref()),
        );
        let added_invite_response_recipients = register_pubkeys(
            &mut self.invite_response_recipients_by_subid,
            &mut self.invite_response_recipient_ref_counts,
            subid,
            invite_response_subscription_recipients(subid, filter_json.as_ref()),
        );

        RuntimeSubscriptionRegistration {
            added_app_keys_authors,
            added_message_authors,
            added_invite_response_recipients,
        }
    }

    pub fn unregister_subscription(&mut self, subid: impl AsRef<str>) {
        let subid = subid.as_ref().trim();
        if subid.is_empty() {
            return;
        }

        unregister_pubkeys(
            &mut self.app_keys_authors_by_subid,
            &mut self.app_keys_author_ref_counts,
            subid,
        );
        unregister_pubkeys(
            &mut self.message_authors_by_subid,
            &mut self.message_author_ref_counts,
            subid,
        );
        unregister_pubkeys(
            &mut self.invite_response_recipients_by_subid,
            &mut self.invite_response_recipient_ref_counts,
            subid,
        );
    }

    pub fn tracked_app_keys_authors(&self) -> Vec<PublicKey> {
        sorted_pubkeys(&self.app_keys_author_ref_counts)
    }

    pub fn tracked_message_authors(&self) -> Vec<PublicKey> {
        sorted_pubkeys(&self.message_author_ref_counts)
    }

    pub fn tracked_invite_response_recipients(&self) -> Vec<PublicKey> {
        sorted_pubkeys(&self.invite_response_recipient_ref_counts)
    }
}

pub fn build_direct_message_backfill_filter(
    authors: impl IntoIterator<Item = PublicKey>,
    since_seconds: u64,
    limit: usize,
) -> Filter {
    let mut unique_authors = Vec::new();
    let mut seen_authors = HashSet::new();
    for author in authors {
        if seen_authors.insert(author) {
            unique_authors.push(author);
        }
    }

    Filter::new()
        .kind(Kind::from(MESSAGE_EVENT_KIND as u16))
        .authors(unique_authors)
        .since(Timestamp::from(since_seconds))
        .limit(limit)
}

pub fn build_app_keys_backfill_filter(
    authors: impl IntoIterator<Item = PublicKey>,
    since_seconds: u64,
    limit: usize,
) -> Filter {
    let mut unique_authors = Vec::new();
    let mut seen_authors = HashSet::new();
    for author in authors {
        if seen_authors.insert(author) {
            unique_authors.push(author);
        }
    }

    Filter::new()
        .kind(Kind::from(APP_KEYS_EVENT_KIND as u16))
        .authors(unique_authors)
        .since(Timestamp::from(since_seconds))
        .limit(limit)
}

pub fn build_invite_response_backfill_filter(
    recipients: impl IntoIterator<Item = PublicKey>,
    since_seconds: u64,
    limit: usize,
) -> Filter {
    let mut unique_recipients = Vec::new();
    let mut seen_recipients = HashSet::new();
    for recipient in recipients {
        if seen_recipients.insert(recipient) {
            unique_recipients.push(recipient.to_hex());
        }
    }

    Filter::new()
        .kind(Kind::from(INVITE_RESPONSE_KIND as u16))
        .custom_tags(SingleLetterTag::lowercase(Alphabet::P), unique_recipients)
        .since(Timestamp::from(since_seconds))
        .limit(limit)
}

pub fn build_runtime_backfill_filters(
    registration: &RuntimeSubscriptionRegistration,
    since_seconds: u64,
    limit: usize,
) -> Vec<Filter> {
    let mut filters = Vec::new();
    if !registration.added_app_keys_authors.is_empty() {
        filters.push(build_app_keys_backfill_filter(
            registration.added_app_keys_authors.iter().copied(),
            since_seconds,
            limit,
        ));
    }
    if !registration.added_message_authors.is_empty() {
        filters.push(build_direct_message_backfill_filter(
            registration.added_message_authors.iter().copied(),
            since_seconds,
            limit,
        ));
    }
    if !registration.added_invite_response_recipients.is_empty() {
        filters.push(build_invite_response_backfill_filter(
            registration
                .added_invite_response_recipients
                .iter()
                .copied(),
            since_seconds,
            limit,
        ));
    }
    filters
}

pub fn app_keys_subscription_authors(
    subid: impl AsRef<str>,
    filter_json: impl AsRef<str>,
) -> Vec<PublicKey> {
    let subid = subid.as_ref().trim();
    if subid.is_empty() {
        return Vec::new();
    }

    let Ok(decoded) = serde_json::from_str::<Value>(filter_json.as_ref()) else {
        return Vec::new();
    };
    let Some(decoded_filter) = decoded.as_object() else {
        return Vec::new();
    };

    if !has_kind(decoded_filter, APP_KEYS_EVENT_KIND) {
        return Vec::new();
    }

    parse_pubkey_values(decoded_filter.get("authors"))
}

pub fn direct_message_subscription_authors(
    subid: impl AsRef<str>,
    filter_json: impl AsRef<str>,
) -> Vec<PublicKey> {
    let subid = subid.as_ref().trim();
    if !is_direct_message_subscription_id(subid) {
        return Vec::new();
    }

    let Ok(decoded) = serde_json::from_str::<Value>(filter_json.as_ref()) else {
        return Vec::new();
    };
    let Some(decoded_filter) = decoded.as_object() else {
        return Vec::new();
    };

    if !has_kind(decoded_filter, MESSAGE_EVENT_KIND) {
        return Vec::new();
    }

    parse_pubkey_values(decoded_filter.get("authors"))
}

pub fn invite_response_subscription_recipients(
    subid: impl AsRef<str>,
    filter_json: impl AsRef<str>,
) -> Vec<PublicKey> {
    let subid = subid.as_ref().trim();
    if !is_invite_response_subscription_id(subid) {
        return Vec::new();
    }

    let Ok(decoded) = serde_json::from_str::<Value>(filter_json.as_ref()) else {
        return Vec::new();
    };
    let Some(decoded_filter) = decoded.as_object() else {
        return Vec::new();
    };

    if !has_kind(decoded_filter, INVITE_RESPONSE_KIND) {
        return Vec::new();
    }

    parse_pubkey_values(decoded_filter.get("#p"))
}

fn has_kind(decoded_filter: &serde_json::Map<String, Value>, kind: u32) -> bool {
    decoded_filter
        .get("kinds")
        .and_then(Value::as_array)
        .map(|kinds| {
            kinds
                .iter()
                .any(|value| value.as_u64() == Some(kind as u64))
        })
        .unwrap_or(false)
}

fn parse_pubkey_values(values: Option<&Value>) -> Vec<PublicKey> {
    let Some(values) = values.and_then(Value::as_array) else {
        return Vec::new();
    };

    let mut parsed = Vec::new();
    let mut seen = HashSet::new();
    for value in values {
        let Some(pubkey_hex) = value.as_str() else {
            continue;
        };
        let normalized = pubkey_hex.trim().to_lowercase();
        if normalized.len() != 64 {
            continue;
        }
        let Ok(pubkey) = pubkey_from_hex(&normalized) else {
            continue;
        };
        if seen.insert(pubkey) {
            parsed.push(pubkey);
        }
    }
    parsed
}

fn register_pubkeys(
    pubkeys_by_subid: &mut HashMap<String, Vec<PublicKey>>,
    pubkey_ref_counts: &mut HashMap<PublicKey, usize>,
    subid: &str,
    pubkeys: Vec<PublicKey>,
) -> Vec<PublicKey> {
    if pubkeys.is_empty() {
        return Vec::new();
    }

    let mut added = Vec::new();
    for pubkey in &pubkeys {
        let ref_count = pubkey_ref_counts.entry(*pubkey).or_insert(0);
        if *ref_count == 0 {
            added.push(*pubkey);
        }
        *ref_count += 1;
    }
    pubkeys_by_subid.insert(subid.to_string(), pubkeys);
    added.sort_by_key(|pubkey| pubkey.to_hex());
    added
}

fn unregister_pubkeys(
    pubkeys_by_subid: &mut HashMap<String, Vec<PublicKey>>,
    pubkey_ref_counts: &mut HashMap<PublicKey, usize>,
    subid: &str,
) {
    let Some(previous_pubkeys) = pubkeys_by_subid.remove(subid) else {
        return;
    };

    for pubkey in previous_pubkeys {
        let next_count = pubkey_ref_counts
            .get(&pubkey)
            .copied()
            .unwrap_or(1)
            .saturating_sub(1);
        if next_count == 0 {
            pubkey_ref_counts.remove(&pubkey);
        } else {
            pubkey_ref_counts.insert(pubkey, next_count);
        }
    }
}

fn sorted_pubkeys(pubkey_ref_counts: &HashMap<PublicKey, usize>) -> Vec<PublicKey> {
    let mut pubkeys: Vec<PublicKey> = pubkey_ref_counts.keys().copied().collect();
    pubkeys.sort_by_key(|pubkey| pubkey.to_hex());
    pubkeys
}

fn is_direct_message_subscription_id(subid: &str) -> bool {
    !subid.is_empty()
        && (subid == DIRECT_MESSAGE_RUNTIME_SUBSCRIPTION_ID
            || subid.starts_with(DIRECT_MESSAGE_RUNTIME_SUBSCRIPTION_PREFIX)
            || subid == DIRECT_MESSAGE_PROTOCOL_SUBSCRIPTION_ID
            || subid.starts_with(DIRECT_MESSAGE_PROTOCOL_SUBSCRIPTION_PREFIX)
            || subid.starts_with(SESSION_CURRENT_SUBSCRIPTION_PREFIX)
            || subid.starts_with(SESSION_NEXT_SUBSCRIPTION_PREFIX))
}

fn is_invite_response_subscription_id(subid: &str) -> bool {
    !subid.is_empty() && subid.starts_with(INVITE_RESPONSE_SUBSCRIPTION_PREFIX)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn pubkey(hex: &str) -> PublicKey {
        PublicKey::parse(hex).unwrap()
    }

    #[test]
    fn parses_app_keys_subscription_authors() {
        let authors = app_keys_subscription_authors(
            "ndr-protocol",
            r##"{
                "kinds":[37368],
                "authors":[
                    "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
                    "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
                    "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
                    "nope"
                ]
            }"##,
        );

        assert_eq!(
            authors,
            vec![
                pubkey("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"),
                pubkey("bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"),
            ]
        );
        assert!(app_keys_subscription_authors(
            "ndr-protocol",
            r##"{"kinds":[7368],"authors":["aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"]}"##,
        )
        .is_empty());
        assert!(app_keys_subscription_authors(
            "",
            r##"{"kinds":[37368],"authors":["aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"]}"##,
        )
        .is_empty());
    }

    #[test]
    fn parses_invite_response_subscription_recipients() {
        let recipients = invite_response_subscription_recipients(
            "invite-responses-a",
            r##"{
                "kinds":[1059],
                "#p":[
                    "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
                    "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
                    "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
                    "nope"
                ]
            }"##,
        );

        assert_eq!(
            recipients,
            vec![
                pubkey("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"),
                pubkey("bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"),
            ]
        );
        assert!(invite_response_subscription_recipients(
            "ndr-runtime-messages",
            r##"{"kinds":[1059],"#p":["aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"]}"##,
        )
        .is_empty());
    }

    #[test]
    fn tracks_runtime_message_authors_and_invite_response_recipients() {
        let mut tracker = RuntimeSubscriptionTracker::new();

        let first = tracker.register_subscription(
            "ndr-protocol",
            r#"{"kinds":[37368],"authors":["aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"]}"#,
        );
        assert_eq!(
            first.added_app_keys_authors,
            vec![pubkey(
                "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
            )]
        );
        assert!(first.added_message_authors.is_empty());
        assert!(first.added_invite_response_recipients.is_empty());

        let second = tracker.register_subscription(
            "ndr-runtime-messages",
            r#"{"kinds":[1060],"authors":["cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc"]}"#,
        );
        assert!(second.added_app_keys_authors.is_empty());
        assert_eq!(
            second.added_message_authors,
            vec![pubkey(
                "cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc"
            )]
        );
        assert!(second.added_invite_response_recipients.is_empty());

        let third = tracker.register_subscription(
            "invite-responses-b",
            r##"{"kinds":[1059],"#p":["bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"]}"##,
        );
        assert!(third.added_app_keys_authors.is_empty());
        assert!(third.added_message_authors.is_empty());
        assert_eq!(
            third.added_invite_response_recipients,
            vec![pubkey(
                "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"
            )]
        );
        assert_eq!(
            tracker.tracked_app_keys_authors(),
            vec![pubkey(
                "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
            )]
        );
        assert_eq!(
            tracker.tracked_message_authors(),
            vec![pubkey(
                "cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc"
            )]
        );
        assert_eq!(
            tracker.tracked_invite_response_recipients(),
            vec![pubkey(
                "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"
            )]
        );

        let fourth = tracker.register_subscription(
            "invite-responses-c",
            r##"{"kinds":[1059],"#p":["bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"]}"##,
        );
        assert!(fourth.added_invite_response_recipients.is_empty());

        tracker.unregister_subscription("invite-responses-b");
        assert_eq!(
            tracker.tracked_invite_response_recipients(),
            vec![pubkey(
                "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"
            )]
        );
        tracker.unregister_subscription("invite-responses-c");
        assert!(tracker.tracked_invite_response_recipients().is_empty());
    }

    #[test]
    fn builds_runtime_backfill_filters() {
        let registration = RuntimeSubscriptionRegistration {
            added_app_keys_authors: vec![pubkey(
                "cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc",
            )],
            added_message_authors: vec![pubkey(
                "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
            )],
            added_invite_response_recipients: vec![pubkey(
                "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
            )],
        };

        let filters = build_runtime_backfill_filters(&registration, 1234, 50);
        assert_eq!(filters.len(), 3);
        let app_keys = serde_json::to_value(&filters[0]).unwrap();
        assert_eq!(app_keys["kinds"], serde_json::json!([37368]));
        assert_eq!(
            app_keys["authors"],
            serde_json::json!(["cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc"])
        );
        assert_eq!(app_keys["since"], serde_json::json!(1234));
        assert_eq!(app_keys["limit"], serde_json::json!(50));

        let direct = serde_json::to_value(&filters[1]).unwrap();
        assert_eq!(direct["kinds"], serde_json::json!([1060]));
        assert_eq!(
            direct["authors"],
            serde_json::json!(["aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"])
        );
        assert_eq!(direct["since"], serde_json::json!(1234));
        assert_eq!(direct["limit"], serde_json::json!(50));

        let invite_response = serde_json::to_value(&filters[2]).unwrap();
        assert_eq!(invite_response["kinds"], serde_json::json!([1059]));
        assert_eq!(
            invite_response["#p"],
            serde_json::json!(["bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"])
        );
        assert_eq!(invite_response["since"], serde_json::json!(1234));
        assert_eq!(invite_response["limit"], serde_json::json!(50));
    }
}
